Channel assignment and general improvements

schote opened the pull request:
### Channel Assignment

Added a new attribute `channel_assignment` of type `Dimensions` to `AcquisitionParameter`.
The attribute takes the integer values `1, 2 and 3` to denote the spectrum card output channel for the gradient channels `x, y, and z`. A check was added to the `SequenceProvider` which ensures, that each value is set only once (i.e. a one to one mapping of channel integers and gradients is ensured).

### Acquisition Parameter

- For saving the acquisition parameters npw json format is used instead of pickle
- `AcquisitionParameters` and `Dimension` dataclasses are defined as "normal" dataclasses, i.e. not frozen or hashable. The comparison of instances is now done by using the dictionary representation.
- Changes of a child with type `Dimensions` now also triggers a call to save the acquisition parameters, i.e. `parameter.gradient_offset.x = 0` now also updates the acquisition parameters on the disk. Since this was mission, a frozen dataclass was used previously to ensure, that the attribute update is recognized.

### SequenceProvider

- Improved the overall structure
- Implemented separate functions for validations (`check_amplitude`, `check_sequence`, `check_parameter`)
- Gradient waveforms are organized in the unrolled sequence dependent on the channel assignment

ckolbPTB commented on src/console/interfaces/acquisition_parameter.py:
```diff
@@ -11,14 +11,23 @@
+    return Dimensions(x=0, y=0, z=0)
 
 
 def _fov_factory() -> Dimensions:
```
```suggestion
def _fov_scaling_factory() -> Dimensions:
```

schote commented:
> I left a few comments on general code style. I don't know enough about the console to be able to critically review this code but my suggestion would be to add further tests for sequence_provider.py to also check all the exceptions are correctly raised.
> 
> As this is a general code improvement PR maybe also add a test for FID.
> 
> I understand that testing the spcm-control part is very difficult without actually using it with the cards. Maybe using something like monkeypatch might allow you to at least have some basic tests. Otherwise improving/extending it in the future will be very challenging.

Thank you @ckolbPTB for your feedback and the suggestions, I added some additional tests for the sequence provider class and addressed your comments.

Regarding the test for the FID sequence, you though about something similar as we already have in place for the SE based sequence?

Unfortunately, testing the spectrum-card–related code is not straightforward. A small simulator is available for the measurement cards (within the [spectrum control center](https://spectrum-instrumentation.com/products/drivers_examples/spectrum_control_center.php)), and integrating it into the CI pipeline could help increase coverage of the relevant code.

h3lg3 commented on tests/sequence_tests/test_sequence_provider.py:
```diff
@@ -14,26 +20,262 @@ def test_unrolling(seq_provider, test_sequence, acquisition_parameter):
+    test_sequence,
+    acquisition_parameter,
+):
+    """unroll_sequence should fail when Larmor frequency violates Nyquist limit."""
```
This docstring is incorrect

h3lg3 commented on src/console/pulseq_interpreter/sequence_provider.py:
```diff
@@ -723,3 +516,211 @@ def plot_unrolled(
+            gradient amplitude exceeds channel maximum output level
+        """
+        try:
+            # Calculate gradient waveform scaling, substract gain and efficiency index by 1,
```
typo "substract" -> subtract

h3lg3 commented on src/console/pulseq_interpreter/sequence_provider.py:
```diff
@@ -723,3 +516,211 @@ def plot_unrolled(
+        """
+        try:
+            # Calculate gradient waveform scaling, substract gain and efficiency index by 1,
+            # because these lists do not include the RF channel (i.e. gradient channe 1 corresponds to index 0)
```
typo "gradient channe"

h3lg3 commented on src/console/interfaces/acquisition_data.py:
```diff
@@ -59,7 +59,7 @@ def __post_init__(self) -> None:
                 "time": datetime_now.strftime("%H:%M:%S"),
```
Double typo on top: "Dictionarz" and below "(ISMR)MRD"
